src/v/bytes/details/io_fragment.h

```diff
@@ -38,7 +42,11 @@ class io_fragment {
     io_fragment& operator=(io_fragment&& o) noexcept = delete;
     io_fragment(const io_fragment& o) = delete;
     io_fragment& operator=(const io_fragment& o) = delete;
-    ~io_fragment() noexcept = default;
+    ~io_fragment() noexcept {
+        vassert(
+          !_next && !_prev,
+          "io_fragment cleanup failure and possible memory leak");
```

Comment:
Adding this check in adds ~3% to the iceberg benchmark. I'm happy to make this `#ifndef NDEBUG` if folks think that's a good tradeoff.

Reply:
> I'm happy to make this #ifndef NDEBUG if folks think that's a good tradeoff.

I'd vote for doing that yeah.

Reply:
Or introduce `dassert` which is exactly `vassert` but only in debug mode. I've wanted that several times.

Reply:
Love it. Will push something up soon